src/controller/artifact/processor/wasm/wasm.go

```diff
@@ -0,0 +1,130 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package wasm
+
+import (
+	"context"
+	"encoding/json"
+	"github.com/goharbor/harbor/src/controller/artifact/processor"
+	"github.com/goharbor/harbor/src/controller/artifact/processor/base"
+	"github.com/goharbor/harbor/src/lib/errors"
+	"github.com/goharbor/harbor/src/lib/log"
+	"github.com/goharbor/harbor/src/pkg/artifact"
+	"github.com/opencontainers/image-spec/specs-go/v1"
+)
+
+// const definitions
+const (
+	// ArtifactTypeImage is the artifact type for image
+	ArtifactTypeWASM        = "WASM"
+	AdditionTypeBuildHistory = "BUILD_HISTORY"
+	mediaType        = "application/vnd.wasm.config.v1+json"
+)
+
+func init() {
+	pc := &manifestV2Processor{}
+	pc.ManifestProcessor = base.NewManifestProcessor()
+	mediaTypes := []string{
+		mediaType,
+	}
+	if err := processor.Register(pc, mediaTypes...); err != nil {
+		log.Errorf("failed to register processor for media type %v: %v", mediaTypes, err)
+		return
+	}
+}
+
+// manifestV2Processor processes image with OCI manifest and docker v2 manifest
+type manifestV2Processor struct {
+	*base.ManifestProcessor
+}
+
+func (m *manifestV2Processor) AbstractMetadata(ctx context.Context, art *artifact.Artifact, manifest_body []byte) error {
+
+	art.ExtraAttrs = map[string]interface{}{}
+	art.ExtraAttrs["ArtifactType"] = "WebAssembly"
+
+	manifest := &v1.Manifest{}
+	if err := json.Unmarshal(manifest_body, manifest); err != nil {
+		return err
+	}
+
+	if manifest.Annotations["module.wasm.image/variant"]=="compat" || manifest.Annotations["run.oci.handler"]=="wasm" {
+
+		// for annotation way
+		config := &v1.Image{}
+		if err := m.UnmarshalConfig(ctx, art.RepositoryName, manifest_body, config); err != nil {
+			return err
+		}
+		art.ExtraAttrs["manifest.config.mediaType"] = manifest.Config.MediaType
+		if art.ExtraAttrs == nil {
+			art.ExtraAttrs = map[string]interface{}{}
+		}
+		art.ExtraAttrs["created"] = config.Created
+		art.ExtraAttrs["architecture"] = config.Architecture
+		art.ExtraAttrs["os"] = config.OS
+		art.ExtraAttrs["config"] = config.Config
+		// if the author is null, try to get it from labels:
+		// https://docs.docker.com/engine/reference/builder/#maintainer-deprecated
+		author := config.Author
+		if len(author) == 0 && len(config.Config.Labels) > 0 {
+			author = config.Config.Labels["maintainer"]
+		}
+		art.ExtraAttrs["author"] = author
+	}else {
+
+		// for wasm-to-oci way
+		art.ExtraAttrs["manifest.config.mediaType"] = mediaType
+		art.ExtraAttrs["manifest.layers.mediaType"] = manifest.Layers[0].MediaType
+		art.ExtraAttrs["org.opencontainers.image.title"] = manifest.Layers[0].Annotations["org.opencontainers.image.title"]
+	}
+	return nil
+}
+
+func (m *manifestV2Processor) AbstractAddition(ctx context.Context, artifact *artifact.Artifact, addition string) (*processor.Addition, error) {
+
+	if addition != AdditionTypeBuildHistory {
```

Comment:
Similar question, the WebAssembly also has the build history concept, right?

Reply:
yes, it happens in annotation fashion wasm artifact